src/main/java/net/openhft/chronicle/queue/reader/ChronicleReader.java

```diff
@@ -191,23 +171,50 @@ public void execute() {
 
     }
 
-    private boolean matchLimitReached(long matches) {
-        return matchLimit > 0 && matches >= matchLimit;
+    /**
+     * Configure the content-based limiter if it was specified
+     */
+    private void configureContentBasedLimiter() {
+        if (contentBasedLimiter != null) {
+            contentBasedLimiter.configure(this);
+        }
     }
 
-    public boolean readOne(@NotNull QueueEntryHandler messageConverter, @NotNull ExcerptTailer tailer, @NotNull Consumer<String> messageConsumer) {
-        try (DocumentContext dc = pollMethod.apply(tailer)) {
-            if (!dc.isPresent()) {
-                return false;
+    /**
+     * Check if the content-based limit has been reached
+     *
+     * @param tailer The Tailer we're using to read the queue
+     * @return true if we should halt reading, false otherwise
+     */
+    private boolean shouldHaltReadingDueToContentBasedLimit(ExcerptTailer tailer) {
+        if (contentBasedLimiter == null) {
+            return false;
+        }
+        long originalIndex = tailer.index();
+        try (final DocumentContext documentContext = tailer.readingDocument()) {
+            if (documentContext.isPresent()) {
+                return contentBasedLimiter.shouldHaltReading(documentContext);
             }
+            return false;
+        } finally {
+            tailer.moveToIndex(originalIndex);
+        }
+    }
 
+    private QueueEntryReader createQueueEntryReader(ExcerptTailer tailer, MessageConsumer messageConsumer) {
+        if (methodReaderInterface == null) {
             if (customPlugin == null) {
-                messageConverter.accept(dc.wire(), messageConsumer);
+                return new VanillaQueueEntryReader(tailer, pollMethod, entryHandlerFactory.get(), messageConsumer);
             } else {
-                customPlugin.onReadDocument(dc, messageConsumer);
+                return new CustomPluginQueueEntryReader(tailer, pollMethod, customPlugin, messageConsumer);
             }
+        } else {
+            return new MethodReaderQueueEntryReader(tailer, messageConsumer, wireType, methodReaderInterface, showMessageHistory);
         }
-        return true;
+    }
```

Comment:
There are essentially 3 different ways of turning entries from the queue into strings, they're "vanilla", customPlugin and methodReader... I think having the separate `QueueEntryReaders` for each type is nice because it makes it clear which settings are for which type. It's a bit ugly because there are subtle differences between the three approaches. Hopefully this structure makes it easier to identify functionality we don't need.